### PR TITLE
Build DMD with -dip25

### PR DESCRIPTION
**Pull request by WalterBright**
Add `-transition=noDIP25` to turn it off.

### CODE REVIEW

**Comment by rainers** [comment]
Looking at the failures in dmd code, I don't see much benefit in adding annotations that are ignored later anyway. Should enforcement be limited to `@safe` code?

File: changelog/DIP25-is-default.dd
Comment (wilzbach) [inline changelog/DIP25-is-default.dd]: I'm not sure where the doc system can handle entries without a title.

Comment (thewilsonator) [reply changelog/DIP25-is-default.dd]: It can't. Needs to be a title followed by a blank line followed by a description, I learnt that the hard way.

Comment (wilzbach) [reply changelog/DIP25-is-default.dd]: I know, I wrote it. I was just trying to be gentle.

Comment (thewilsonator) [reply changelog/DIP25-is-default.dd]: and I'm removing any ambiguity because I've had about as much fun with these things as Walter and I know I'd rather just be told what needs doing ;-)

**Review by wilzbach — DISMISSED** [review]
This can't be merged before Buildkite is green and at the moment it fails at the build stage because it can't build dub.

**Comment by Geod24** [comment]
I'm with @jacob-carlborg here, if it's not a deprecation, it's going to be fairly bad, even if there is a way to disable it.

**Comment by WalterBright** [comment]
CyberShadow/DAutoTest fails with:
```
__stdin.d(4): Error: returning `&this.numdg` escapes a reference to parameter `this`, perhaps annotate with `return`
```
Anyone know where this file is?

**Comment by ZombineDev** [comment]
@WalterBright The errors produced by compiling certain examples are from the language specification that are wrapped with the `SPEC_RUNNABLE_EXAMPLE_COMPILE` macro. Here are the relevant locations:
* https://github.com/dlang/dlang.org/blob/v2.084.0-rc.1/spec/function.dd#L599-L635
* https://github.com/dlang/dlang.org/blob/v2.084.0-rc.1/spec/struct.dd#L1467-L1486

**Comment by WalterBright** [comment]
@ZombineDev Thanks! https://github.com/dlang/dlang.org/pull/2533

**Comment by thewilsonator** [comment]
See https://github.com/dlang/dub/pull/1632

**Comment by WalterBright** [comment]
Blocked by dub failures:
```
source/dub/internal/vibecompat/data/json.d(152): Error: returning `this` escapes a reference to parameter `this`, perhaps annotate with `return`
```

**Comment by thewilsonator** [comment]
Done.

**Comment by thewilsonator** [comment]
Another one https://github.com/CyberShadow/DustMite/pull/56

**Review by jacob-carlborg — DISMISSED** [review]
This need to go through the regular deprecation process. All the build failures clearly show that as well.

**Comment by wilzbach** [comment]
dlang/tools PR: https://github.com/dlang/tools/pull/360

**Comment by wilzbach** [comment]
This still looks pretty bad:

image

About half of the D ecosystem tested would break with this PR.

**Comment by wilzbach** [comment]
Also note that buildkite dlang/phobos fails because it uses LATEST as test compiler (this means also its dmd is build with latest) and thus we get this error when building DMD:

```

dmd/backend/dvarstats.d(272): Error: returning `&this.sortedSymtab` escapes a reference to parameter `this`, perhaps annotate with `return`
  | posix.mak:571: recipe for target '../generated/linux/release/64/dvarstats.o' failed
```

It doesn't appear on the auto-tester as there 2.079 is used for testing.

**Comment by jacob-carlborg** [comment]
> About half of the D ecosystem tested would break with this PR.

It's obvious we cannot merge this PR as is.

**Comment by ZombineDev** [comment]
https://github.com/weka-io/mecca/pull/13 fixes `-dip25` for mecca.

**Comment by thewilsonator** [comment]
>Also note that buildkite dlang/phobos fails because it uses LATEST as test compile

It seems that the changes need to be split between updating the dmd codebase to be dip25 compliant and the change of default.

**Comment by Geod24** [comment]
Well it needs to be made a deprecation first.

The easy path is to make the deprecation about not using the `-dip25` switch, instead of deprecating places where `dip25` would trigger. But making it the default, and triggering an error, would be terrible for the users.

**Comment by thewilsonator** [comment]
> would be terrible for the users.

I talking about making dmd _itself_ dip25 compliant (as a requisite for building itself deprecation free).

**Comment by WalterBright** [comment]
> Also note that buildkite dlang/phobos fails

Are you sure? dvarstats.d looks like it was corrected.

**Comment by thewilsonator** [comment]
@WalterBright 

> Also note that buildkite dlang/phobos fails because it uses LATEST as test compiler

i.e. the changes need to be staggered first make dmd dip25 compliant, then flip the deprecation switch.

**Comment by wilzbach** [comment]
Added a commit which disables the `-dip25` default flip for now, s.t. these changes can already be merged.

**Comment by wilzbach** [comment]
Also I opened up a new PR for just changing the default for `-dip25`, s.t. this change doesn't get forgotten after this PR has been merged -> https://github.com/dlang/dmd/pull/9194

**Review by thewilsonator — APPROVED** [review]

**Comment by wilzbach** [comment]
tl;dr: this PR no longer sets `-dip25` as default, but just allows building DMD (and its testsuite) with `-dip25` (though to be pedantic for now `-dip25` isn't added by the default to the testsuite, just when building dmd).

Imho, a proper deprecation cycle still needs to follow (i.e. the errors resulting from activating `-dip25` need to be deprecations for a few releases), but this discussion can be done in https://github.com/dlang/dmd/pull/9194 as this PR now only contains the non-controversial changes.

**Pull request merged**
